crypto/threads_win.c

```diff
@@ -0,0 +1,162 @@
+ *    nor may "OpenSSL" appear in their names without prior written
+ *    permission of the OpenSSL Project.
+ *
+ * 6. Redistributions of any form whatsoever must retain the following
+ *    acknowledgment:
+ *    "This product includes software developed by the OpenSSL Project
+ *    for use in the OpenSSL Toolkit (http://www.openssl.org/)"
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE OpenSSL PROJECT ``AS IS'' AND ANY
+ * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
+ * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ====================================================================
+ */
+
+#include "e_os.h"
+
+#include <openssl/crypto.h>
+
+#if defined(OPENSSL_THREADS) && !defined(CRYPTO_TDEBUG) && defined(OPENSSL_SYS_WINDOWS)
+
+int CRYPTO_MUTEX_init(CRYPTO_MUTEX *lock)
+{
+    if (InitializeCriticalSectionAndSpinCount((CRITICAL_SECTION *) lock, 0x400)
+          != 0)
+        return 0;
+
+    return 1;
+}
+
+int CRYPTO_MUTEX_lock_read(CRYPTO_MUTEX *lock)
+{
+    EnterCriticalSection((CRITICAL_SECTION *) lock);
+    return 1;
+}
+
+int CRYPTO_MUTEX_lock_write(CRYPTO_MUTEX *lock)
+{
+    EnterCriticalSection((CRITICAL_SECTION *) lock);
+    return 1;
+}
+
+int CRYPTO_MUTEX_unlock(CRYPTO_MUTEX *lock)
+{
+    LeaveCriticalSection((CRITICAL_SECTION *) lock);
+    return 1;
+}
+
+void CRYPTO_MUTEX_cleanup(CRYPTO_MUTEX *lock)
+{
+    DeleteCriticalSection((CRITICAL_SECTION *) lock);
+}
+
+# if _WIN32_WINNT < 0x0600
+
+#  define ONCE_UNINITED     0
+#  define ONCE_ININIT       1
+#  define ONCE_DONE         2
+
+void CRYPTO_ONCE_run(CRYPTO_ONCE *once, void (*init)(void))
+{
+    LONG volatile *lock = (LONG *)once;
+    LONG result;
+
+    if (*lock == ONCE_DONE)
+        return;
+
+    do {
+        result = InterlockedCompareExchange(lock, ONCE_ININIT, ONCE_UNINITED);
+        if (result == ONCE_UNINITED) {
+            init();
+            *lock = ONCE_DONE;
+            return;
+        }
+    } while (result == ONCE_ININIT);
+}
+
+# else
+
+BOOL CALLBACK once_cb(PINIT_ONCE once, PVOID p, PVOID *pp)
+{
+    void (*init)(void) = p;
+
+    init();
+
+    return TRUE;
+}
+
+void CRYPTO_ONCE_run(CRYPTO_ONCE *once, void (*init)(void))
+{
+    InitOnceExecuteOnce((INIT_ONCE *) once, once_cb, init, NULL);
+}
+
+# endif
+
+int CRYPTO_THREAD_LOCAL_init(CRYPTO_THREAD_LOCAL *key, void (*cleanup)(void *))
+{
+    *key = TlsAlloc();
```

Comment:
Do we want to try to avoid writing to *key in the failure case?  (The no-threads version does so, and the pthreads one is subject to the pthread implementation.)